src/filebrowser/buttons.ts

```diff
@@ -408,12 +453,35 @@ namespace Private {
    * Create a new folder.
    */
   export
-  function createNewFolder(widget: FileButtons): void {
-    widget.model.newUntitled({ type: 'directory' }).then(contents => {
-      widget.model.refresh();
+  function createNewFolder(widget: FileButtons): Promise<void> {
+    let dir = widget.model.newUntitled({ type: 'directory' });
+
+    dir.then(contents => {
+      Private.setDirName(contents.name);
     }).catch(error => {
       utils.showErrorMessage(widget, 'New Folder Error', error);
     });
+    return widget.model.refresh();
+  }
+
+  /**
+   * Calls createNewFolder and selects that folder to be renamed.
+   */
+  export
+  function createAndRename(widget: FileButtons): void {
+    let prom = createNewFolder(widget);
+
+    prom.then(contents => {
```

Comment:
These two then clauses will be the core logic that you move inside the `fileChanged` handler in the browser